### DISCUSSION TITLE
Is there any way that I can play audio by Avalonia?

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#16932] [Q&A] [answered]
I want to play audio (I just want to play audio files, not limited to a certain format or compatible with all audio formats) in windows platform and Linux DRM.

In the past, I used NAudio to play audio on Windows, but NAudio doesn't seem to support cross-platform. Avalonia's official API doesn't seem to have anything related to audio playback. Is there a third-party package that can solve this problem?

**Comment by stevemonaco** [comment]
I've heard at least one positive mention of SFML (via SFML.NET's SFML.Audio package). I just tested and it adds about 4MB of native dependencies on win-x64. It seems to publish with the graphics and windowing native deps, too, so you might be able to manually trim them out to save another 2MB.

```cs
private void Play_Click(object? sender, RoutedEventArgs e)
{
    var music = new Music("test.mp3");
    music.Play();
}
```

It worked for me with one MP3, but not another one. Both very old files (encoded in the late 1990s).

**Comment by thevortexcloud** [comment, marked as answer]
> Linux DRM.

This is actually fairly easy as long as you have audio configured. If you have an uncompressed audio file you can just send it to `aplay`.

https://en.wikipedia.org/wiki/Aplay

 If you need something more advanced and cross platform, there is an official VLC binding available for .NET (and also Avalonia if you need video).

https://github.com/videolan/libvlcsharp

If you need something more minimal, you can probably write your own bindings for `libmpv`. They have a C# example here, but I have never tried it myself.

https://github.com/mpv-player/mpv-examples/tree/master/libmpv/csharp

If you want something more "API" like with existing .NET bindings, you can use the SDL audio APIs. But it can be kind of complicated.

https://github.com/flibitijibibo/SDL2-CS

**Reply by kyuranger** [reply]
Well, that's great! Thank you so much!

**Reply by thevortexcloud** [reply]
Glad I could help.

**Reply by pablogupi** [reply]
And are there any compatible with macox arm64 and Windows?

**Reply by thevortexcloud** [reply]
Everything I listed except for `aplay` will run on every major desktop platform and CPU architecture. SDL will run on pretty much every somewhat popular computing device invented, including niche operating systems like AmigaOS and Haiku. 

Mac also has a couple command line programs for playing sounds. EG `afplay`. Which I am guessing is just a port of `aplay`. But I can't really give any specific Mac advice as I don't have a lot of experience using them.

**Reply by pablogupi** [reply]
Ok I will check it out. But for example, LibVLCSharp ew compatible with Mac
but not with Arm64. So it does not work on any Apple Silicon processor.

**Reply by thevortexcloud** [reply]
My understanding is that `libvlc` itself is compatible. You can probably just add the binary yourself and it will probably work.

**Comment by Monsalma** [comment]
I tried irrKlang ([https://www.ambiera.com/irrklang/](https://www.ambiera.com/irrklang/)), but only for Windows Desktop and it was pretty straightforward. Now, I didn't try it myself, but the website says Linux is supported as well. The website makes things easy as there are some code examples. However, eventually I gave up on irrKlang because it doesn't support Android.

So I switched to LibVLCSharp. NuGet package is here - [https://www.nuget.org/packages/LibVLCSharp/](https://www.nuget.org/packages/LibVLCSharp/) and main documentation here - [https://code.videolan.org/videolan/LibVLCSharp/-/blob/3.x/docs/home.md](https://code.videolan.org/videolan/LibVLCSharp/-/blob/3.x/docs/home.md). 

I created a demo project which showcases using Avalonia UI and LibVLCSharp to play audio on Windows Desktop and Android. Hopefully it will be useful to you. 
The complete source code is available on GitHub ([https://github.com/Monsalma/Monsalma.AvaloniaUI/tree/main/Monsalma_AvaloniaAudioTest](https://github.com/Monsalma/Monsalma.AvaloniaUI/tree/main/Monsalma_AvaloniaAudioTest)). 
There is a blog post that explains everything in detail ([https://monsalma.net/avalonia-ui-audio-playback-demo/](https://monsalma.net/avalonia-ui-audio-playback-demo/)) so be sure to check it out.

**Comment by sandreas** [comment]
If you would like to have a simpler and more portable solution, you could try [SoundFlow](https://github.com/LSXPrime/SoundFlow). Cross-Platform (WASM is in progress though), very good performance. Can Play and Record.

**Reply by TheShihan** [reply]
> If you would like to have a simpler and more portable solution, you could try [SoundFlow](https://github.com/LSXPrime/SoundFlow). Cross-Platform (WASM is in progress though), very good performance. Can Play and Record.

Is there something special needed to get it working in Avalonia? I try to play a mp3 file like this but I don't hear any sound:
```
    private static void PlayDummySound()
    {
        using var audioEngine = new MiniAudioEngine(44100, Capability.Playback);

        var player = new SoundPlayer(new StreamDataProvider(File.OpenRead(AppContext.BaseDirectory + @"\Assets\Sounds\test.mp3")));

        Mixer.Master.AddComponent(player);
        player.Play();

        //player.Stop();
        Mixer.Master.RemoveComponent(player);
    }
```

The resource is successfully found, so this cannot be the issue. File can be played normally and heard in Windows.

**Reply by TheShihan** [reply]
OK, nevermind.. i got it working with this code:

```
    private static async Task PlayStreamAsync(Stream soundStream)
    {
        using var audioEngine = new MiniAudioEngine(44100, Capability.Playback);

        var dataProvider = new AssetDataProvider(soundStream);

        var player = new SimpleSoundPlayer(dataProvider);


        player.PlaybackEnded += (sender, args) =>
        {
            Dispatcher.UIThread.InvokeAsync(() => { Mixer.Master.RemoveComponent(player); });
            dataProvider?.Dispose();
        };

        Mixer.Master.AddComponent(player);
        player.Play();

        await Task.Run(() =>
        {
            while (player.State == PlaybackState.Playing)
            {
                Thread.Sleep(100); // Small delay to prevent CPU hogging
            }
        });
    }
 ```

Somehow it needs something to happen between playing and stoping/removal of the player. The example is a console application where Console.ReadKey() is used. But in my avalonia app I had to use something different.

Maybe someone can shine a light onto why this is needed.

**Reply by stevemonaco** [reply]
We used a removal queue along with `OnAudioProcessed` event. You can't remove components while the audio graph is still processing.

That said, you can't publish apps with SoundFlow until the author fixes https://github.com/LSXPrime/SoundFlow/issues/1 which shouldn't be far off.

If you have Accelerate, you can use the `MediaPlayerControl` or `MediaPlayer` (invisible, can run multiple simultaneously though it's not a real mixer) that's included.
